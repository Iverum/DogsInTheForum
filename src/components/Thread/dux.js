// Actions
const ADD_POST = 'ditf/threads/ADD_POST'
const CLEAR_POSTS = 'ditf/threads/CLEAR_POSTS'

// Reducer
export default function reducer(state = {}, action = {}) {
  switch (action.type) {
    case ADD_POST: {
      const currentPosts = state[action.thread] || []
      const newPosts = [...currentPosts, action.data]
      return {
        ...state,
        [action.thread]: newPosts
      }
    }

    case CLEAR_POSTS: {
      return {}
    }

    default: {
      return state
    }
  }
}

// Action creators
export function addPost(thread, post) {
  return {
    type: ADD_POST,
    data: post,
    thread
  }
}

export function clear() {
  return { type: CLEAR_POSTS }
}
