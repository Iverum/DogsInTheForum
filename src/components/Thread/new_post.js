import React, { Component } from 'react';

export default class NewPost extends Component {

  render() {
    return (
      <div className="row">
        <textarea className="u-full-width" placeholder="I cast magic missile into the darkness!"></textarea>
        <input className="button-primary" type="submit" value="Submit" />
      </div>
    )
  }

}
