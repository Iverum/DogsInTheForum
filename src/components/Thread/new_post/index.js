import React, { Component, PropTypes } from 'react'
import moment from 'moment'
import _ from 'lodash'
import handleCommands from './commands'

export default class NewPost extends Component {

  constructor(props) {
    super(props)
    this.handleChange = this.handleChange.bind(this)
    this.submitPost = this.submitPost.bind(this)
    this.state = { value: '' }
  }

  handleChange(event) {
    this.setState({ value: event.target.value })
  }

  submitPost(e) {
    if (this.props.onSubmit) {
      const post = handleCommands(this.state.value)
      post.dice = post.dice.rolledDice
      this.props.onSubmit({
        ...post,
        author: this.props.user.id,
        created: moment().format()
      })
      this.setState({ value: '' })
    }
  }

  render() {
    if (_.isEmpty(this.props.user)) { return null }

    return (
      <div className="row">
        <textarea
          className="u-full-width"
          placeholder="I cast magic missile into the darkness!"
          value={this.state.value}
          onChange={this.handleChange}
        ></textarea>
        <input className="button-primary" type="submit" value="Submit" onClick={this.submitPost} />
      </div>
    )
  }

}

NewPost.propTypes = {
  user: PropTypes.object,
  onSubmit: PropTypes.func
}
