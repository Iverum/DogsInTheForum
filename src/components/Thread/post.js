import React, { Component, PropTypes } from 'react'
import Markdown from 'markdown-it'
import _ from 'lodash'
import cn from 'classnames'
import './thread.css'
import Dice from '../Dice'

const mdRenderer = new Markdown()

export default class Post extends Component {
  renderDice() {
    if (_.isEmpty(this.props.userDice)) { return null }
    const caption = `${this.props.username} rolled:`
    return <Dice dice={this.props.userDice} caption={caption} />
  }

  render() {
    return (
      <section className='row post'>
        <div className='three columns'>
          <img src='https://placehold.it/175x175' />
          <h3 className='username'>{this.props.username}</h3>
        </div>
        <div className='nine columns'>
          <div className={cn({
            row: true,
            bottomBorder: !_.isEmpty(this.props.userDice)
          })}>
            <div className='text' dangerouslySetInnerHTML={{ __html: mdRenderer.render(this.props.text) }} />
          </div>
          <div className='row'>
            {this.renderDice()}
          </div>
        </div>
      </section>
    )
  }

}

Post.defaultProps = {
  userDice: []
}

Post.propTypes = {
  username: PropTypes.string.isRequired,
  text: PropTypes.string.isRequired,
  userDice: PropTypes.arrayOf(PropTypes.shape({
    type: PropTypes.oneOf(['d2', 'd4', 'd6', 'd8', 'd10', 'd12', 'd20']),
    value: PropTypes.number
  }))
}
