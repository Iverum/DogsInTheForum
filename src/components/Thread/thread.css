.right-border {
  border-right: gray solid 1px;
}

.bottom-border {
  border-bottom: gray solid 1px;
}

div.post {
  padding-top: 1rem;
  padding-bottom: 1rem;
}
