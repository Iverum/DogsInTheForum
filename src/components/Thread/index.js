import React, { Component } from 'react'
import { connect } from 'react-redux'
import firebase from 'firebase'
import _ from 'lodash'
import Post from './post'
import NewPost from './new_post'
import * as threadActions from './dux'
import * as userActions from '../User/dux'

export class Thread extends Component {

  constructor(props) {
    super(props)
    this.createNewPost = this.createNewPost.bind(this)
    this.updateThreadFromDatabase = this.updateThreadFromDatabase.bind(this)
  }

  componentWillMount() {
    const uuid = this.props.params.uuid
    this.postRef = firebase.database().ref(`posts/${uuid}`)
    this.postRef.off()
    this.postRef.on('child_added', this.updateThreadFromDatabase)
    this.postRef.on('child_changed', this.updateThreadFromDatabase)
  }

  componentWillUnmount() {
    this.postRef.off()
    this.postRef = null
    this.props.dispatch(threadActions.clear())
  }

  updateThreadFromDatabase(data) {
    const uuid = this.props.params.uuid
    const post = data.val()
    if (!_.has(this.props.users, post.author)) {
      firebase.database().ref(`users/${post.author}`).once('value', data => {
        this.props.dispatch(userActions.addUser(data.val()))
      })
    }
    this.props.dispatch(threadActions.addPost(uuid, post))
  }

  createNewPost(post) {
    // TODO handle post.hand & put the values somewhere
    const { threads, params } = this.props
    const thread = threads[params.uuid]
    let threadLength = 0
    if (thread) {
      threadLength = thread.length
    }
    this.postRef.push(post)
    firebase.database().ref(`threads/${params.uuid}`).child('postCount').set(threadLength + 1)
  }

  renderPosts() {
    const { threads, params } = this.props
    const thread = threads[params.uuid]
    if (!thread || thread.length === 0) { return <p>There are no posts here yet!</p>}
    return thread.map((post, i) => {
      return (
        <Post
          key={i}
          {...post}
          author={this.props.users[post.author]}
        />
      )
    })
  }

  render() {
    return (
      <div className="twelve columns">
        {this.renderPosts()}
        <hr />
        <NewPost
          user={this.props.user}
          onSubmit={this.createNewPost}
        />
      </div>
    )
  }

}

export default connect(state => ({
  threads: state.threads,
  user: state.users.currentUser,
  users: state.users.users
}))(Thread)
