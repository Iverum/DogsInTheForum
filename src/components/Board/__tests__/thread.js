import React from 'react'
import ReactDOM from 'react-dom'
import renderer from 'react-test-renderer'
import Thread from '../thread'

test('thread is rendered correctly', () => {
  const component = renderer.create(
    <Thread
      data={{
        name: 'Test Thread',
        postCount: 0
      }}
    />
  )
  let tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})
