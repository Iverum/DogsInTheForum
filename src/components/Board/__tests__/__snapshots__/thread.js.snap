exports[`test thread is rendered correctly 1`] = `
<section
  className="row">
  <a
    className="six columns"
    onClick={[Function]}
    style={Object {}}>
    <h2
      className="title">
      Test Thread
    </h2>
  </a>
  <span
    className="four columns post-count">
    created by 
  </span>
  <span
    className="two columns post-count">
    0
     posts
  </span>
</section>
`;
